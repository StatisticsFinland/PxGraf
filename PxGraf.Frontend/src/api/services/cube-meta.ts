/* istanbul ignore file */

import ApiClient from "api/client";
import { useQuery } from "react-query";
import { ICubeMeta } from "types/cubeMeta";
import { defaultQueryOptions } from "utils/ApiHelpers";

/**
 * Interface for cube metadata result. Cube is a multi-dimensional result of Px file data filtered by a query.
 * @property {boolean} isLoading - Flag to indicate if the data is still loading.
 * @property {boolean} isError - Flag to indicate if an error occurred during loading.
 * @property {ICubeMeta} data - The cube meta data as an @see {@link ICubeMeta} object.
 */
export interface ICubeMetaResult {
    isLoading: boolean;
    isError: boolean;
    data: ICubeMeta;
}

const fetchCubeMeta = async(idStack: string[]): Promise<ICubeMeta> => {
    const client = new ApiClient();
    const url = `creation/cube-meta/${idStack.join('/')}`;

    return await client.getAsync(url);
}

export const useCubeMetaQuery = (idStack: string[]): ICubeMetaResult => {
    return useQuery(
        ['cube-meta', idStack],
        () => fetchCubeMeta(idStack),
        defaultQueryOptions
    );
}