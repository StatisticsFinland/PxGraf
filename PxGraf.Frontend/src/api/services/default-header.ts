/* istanbul ignore file */

import ApiClient from "api/client";
import { merge } from "lodash";
import { useQuery } from "react-query";
import { MultiLanguageString } from "types/multiLanguageString";
import { Query } from "types/query";
import { buildTableReference, defaultQueryOptions } from "utils/ApiHelpers";

/**
 * Interface for the chart's header result
 * @property {boolean} isLoading - Flag to indicate if the data is still loading.
 * @property {boolean} isError - Flag to indicate if an error occurred while retrieving the header.
 * @property {MultiLanguageString} data - The header data represented as a multi-language string.
 */
export interface IHeaderResult {
    isLoading: boolean;
    isError: boolean;
    data: MultiLanguageString;
}

const fetchDefaultHeader = async(idStack: string[], query: Query): Promise<MultiLanguageString> => {
    const client = new ApiClient();
    const requestBody = JSON.stringify(
        merge(
            { tableReference: buildTableReference(idStack) },
            { variableQueries: query }
        )
    );
  
    const url = 'creation/default-header';
    return await client.postAsync(url, requestBody);
}

export const useDefaultHeaderQuery = (idStack: string[], query: Query): IHeaderResult => {
    return useQuery(
        ['default-header', ...idStack, query],
        () => fetchDefaultHeader(idStack, query),
        {
            ...defaultQueryOptions,
            enabled: query != null,
        }
    );
}