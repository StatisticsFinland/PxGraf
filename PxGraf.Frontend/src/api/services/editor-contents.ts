/* istanbul ignore file */

import ApiClient from "api/client";
import { useQuery } from "react-query";
import { IEditorContentsResponse } from "types/editorContentsResponse";
import { buildCubeQuery, defaultQueryOptions, useDebounceState } from "utils/ApiHelpers";
import { ICubeQuery, Query } from "types/query";

/**
 * Interface for editor contents result. Editor contents are used by the editor to populate the settings for adjusting the visualization.
 * @property {boolean} isLoading - Flag to indicate if the data is still loading.
 * @property {boolean} isError - Flag to indicate if an error occurred during loading.
 * @property {IEditorContentsResponse} data - The editor contents as an @see {@link IEditorContentsResponse} object.
 */ 
export interface IEditorContentsResult {
    isLoading: boolean;
    isError: boolean;
    data: IEditorContentsResponse;
}

const fetchEditorContents = async (idStack: string[], query: Query, metaEdits: ICubeQuery): Promise<IEditorContentsResponse> => {
    const client = new ApiClient();
    const url = 'creation/editor-contents';
    const requestBody = JSON.stringify(buildCubeQuery(query, metaEdits, idStack));
    return await client.postAsync(url, requestBody);
}

export const useEditorContentsQuery = (idStack: string[], query: Query, cubeQuery: ICubeQuery): IEditorContentsResult => {
    [idStack, query, cubeQuery] = useDebounceState(1000, idStack, query, cubeQuery);
    const queryKey = ['editor-contents', ...idStack, query, cubeQuery];
    return useQuery(
        queryKey,
        () => fetchEditorContents(idStack, query, cubeQuery),
        {
            ...defaultQueryOptions,
            enabled: query != null,
            keepPreviousData: true,
        }
    );
}