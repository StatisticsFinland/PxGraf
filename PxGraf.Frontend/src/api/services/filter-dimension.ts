/* istanbul ignore file */

import ApiClient from "api/client";
import { useQuery } from "react-query";
import { IValueFilter, Query } from "types/query";
import { buildTableReference, defaultQueryOptions } from "utils/ApiHelpers";

/**
 * Interface for variable filter result
 * @property {boolean} isLoading - Flag to indicate if the data is still loading.
 * @property {boolean} isError - Flag to indicate if an error occurred during loading.
 * @property {{ [key: string]: string[] }} data - The variable filter data represented by a dictionary. Key is the variable code and value is a list of strings that represent value codes for that variable.
 */
export interface IFilterVariableResult {
    isLoading: boolean;
    isError: boolean;
    data: {[key: string]: string[]}
}

const fetchResolveVariableFilter = async (idStack: string[], variableFilters: { [key: string]: IValueFilter }): Promise<{ [key: string]: string[] }> => {
    const client = new ApiClient();
    const requestBody = JSON.stringify(
        {
            tableReference: buildTableReference(idStack),
            filters: variableFilters,
        }
    );

    const url = 'creation/filter-dimension'
    return await client.postAsync(url, requestBody);
}

export const useResolveVariableFiltersQuery = (idStack: string[], query: Query): IFilterVariableResult => {
    const varQueries = Object.entries(query ?? {});
    const varFilters =  Object.fromEntries(varQueries.map(([variableCode, variableQuery]) => {
        return [variableCode, variableQuery.valueFilter]
    }));

    return useQuery(
        ['filter-dimension', ...idStack, query],
        () => fetchResolveVariableFilter(idStack, varFilters),
        defaultQueryOptions
    );
}