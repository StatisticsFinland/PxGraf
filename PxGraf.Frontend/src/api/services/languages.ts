/* istanbul ignore file */

import ApiClient from "api/client";
import { useQuery } from "react-query";
import { defaultQueryOptions } from "utils/ApiHelpers";

const fetchLanguages = async (): Promise<string[]> => {
    const client = new ApiClient();
    const url = "creation/languages/";

    return await client.getAsync(url);
};

/**
 * Parses a list of languages into a string for display.
 * @param {string[]} languages List of languages.
 * @returns All listed languages in one string, in upper case.
 */
export const parseLanguageString = (languages: string[]): string => {
    return `(${languages.join(", ").toUpperCase()})`;
}

/**
 * Chooses a listing language for the given listing item, trying to prefer the selected uiLanguage.
 * @param {string} id Id of the database or table.
 * @param {string[]} languages Languages supported by the listing item.
 * @param {string} uiLanguage Currently selected UI language.
 * @returns UI language, if it is supported by the listing item's languages. Otherwise the default language of the item.
 */
export const getDatabaseListingLanguage = (id: string, languages: string[], uiLanguage: string): string => {
    return languages.includes(uiLanguage) ? uiLanguage : languages[0];
}

export const useLanguagesQuery = (): string[] => {
    const queryResult = useQuery(
        ['languages'],
        () => fetchLanguages(),
        defaultQueryOptions
    );
    return queryResult.data;
};