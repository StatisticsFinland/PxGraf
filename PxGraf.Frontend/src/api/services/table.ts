/* istanbul ignore file */

import ApiClient from "api/client";
import { useQuery } from "react-query";
import { defaultQueryOptions } from "utils/ApiHelpers";
import { MultiLanguageString } from "../../types/multiLanguageString";

/**
 * Interface for a table list response.
 * @property {string} id - The id of the table.
 * @property {string} type - The type of the table where 't' is a table and 'l' is a list item.
 * @property {string} updated - The last updated date of the table.
 * @property {MultiLanguageString} text - The description of the table as a multi-language string.
 * @property {string[]} languages - Available languages of the table.
 */
export interface ITableListResponse {
    id: string;
    type?: 't' | 'l';
    updated?: string;
    text: MultiLanguageString;
    languages?: string[];
}

/**
 * Interface for a table result.
 * @property {boolean} isLoading - Flag to indicate if the data is still loading.
 * @property {boolean} isError - Flag to indicate if an error occurred during loading.
 * @property {ITableListResponse[]} data - The table list response data.
 */
export interface ITableResult {
    isLoading: boolean;
    isError: boolean;
    data: ITableListResponse[];
}

const fetchTable = async (idStack: string[], lang: string): Promise<ITableListResponse[]> => {
    const client = new ApiClient();
    const path = idStack.join("/");
    const url = "creation/data-bases/" + path;
    const getParams = {
        'lang': lang
    }

    return await client.getAsync(url, getParams);
};

export const sortTableData = (data: ITableListResponse[], primaryLanguage: string, databaseLanguages: string[]): ITableListResponse[] => {
    return [...data].sort((a, b) => {
        const textA = a.text[primaryLanguage] || a.text[databaseLanguages[0]];
        const textB = b.text[primaryLanguage] || b.text[databaseLanguages[0]];

        return textA.localeCompare(textB, primaryLanguage);
    });
};

export const useTableQuery = (idStack: string[], lang: string): ITableResult => {
    // If trying to fetch a database without a language, return empty
    const fetchFunction = lang ?
        () => fetchTable(idStack, lang) :
        () => Promise.resolve([]);

    return useQuery(
        ['table', ...idStack, lang],
        fetchFunction,
        defaultQueryOptions,
    );
};