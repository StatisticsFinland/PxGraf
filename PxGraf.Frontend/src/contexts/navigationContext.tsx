import * as React from "react";

interface INavigationContextData {
    tablePath: string[] | null,
}

interface INavigationContextMethods {
    setTablePath: React.Dispatch<React.SetStateAction<string[] | null>>,
}

type TNavigationContextValue = INavigationContextData & INavigationContextMethods;

interface INavigationProviderProps {
    children: React.ReactNode;
}

const NavigationContext = React.createContext<(TNavigationContextValue)>(null);

export const NavigationProvider: React.FC<INavigationProviderProps> = ({ children }) => {
    const [tablePath, setTablePath] = React.useState<string[] | null>(null);

    const contextValue: TNavigationContextValue = React.useMemo(() => ({
        tablePath,
        setTablePath,
    }), [tablePath]);

    return <NavigationContext.Provider value={contextValue}>{children}</NavigationContext.Provider>
}

export const useNavigationContext = (): TNavigationContextValue => {
    return React.useContext(NavigationContext);
}
