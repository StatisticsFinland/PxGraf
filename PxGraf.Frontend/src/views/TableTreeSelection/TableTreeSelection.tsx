import * as React from 'react';
import { useTranslation } from 'react-i18next';

import { List, ListSubheader, Container } from '@mui/material';
import { NestedList } from 'components/NestedList/NestedList';
import styled from 'styled-components';
import InfoBubble from 'components/InfoBubble/InfoBubble';
import useReplaceQueryParams from 'hooks/useReplaceQueryParams';

const TableTreeSelectionWrapper = styled(Container)`
  padding: 8px;
`;

const SubHeaderWrapper = styled.div`
  display: flex;
`;

/**
 * Table tree selection view. This is the default view of the program where the user can browse available databases and tables and selects a table for visualization.
 * @see {@link NestedList} component is used for fetching and displaying the database and table hierarchy.
 */
export const TableTreeSelection: React.FC = () => {
  const { t } = useTranslation();
  useReplaceQueryParams('/');

  React.useEffect(() => {
      document.title = `${t("pages.tableTreeSelection")} | PxGraf`;
  }, []);

  return (
    <TableTreeSelectionWrapper maxWidth="md">
      <List
        component="nav"
        aria-labelledby="nested-list-subheader"
        subheader={
            <ListSubheader component="div" id="nested-list-subheader">
                <SubHeaderWrapper><div>{t("tableSelect.title")}</div><InfoBubble info={t("infoText.treeSelection")} ariaLabel={t("tableSelect.title")} id="mainContent" /></SubHeaderWrapper>
          </ListSubheader>
        }
      >
        <NestedList path={[]} depth={0}/>
      </List>
    </TableTreeSelectionWrapper>
  );
}

export default TableTreeSelection;