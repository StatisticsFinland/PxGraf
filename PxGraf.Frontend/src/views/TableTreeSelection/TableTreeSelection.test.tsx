import React from 'react';
import { render } from '@testing-library/react';
import { ITableResult } from 'api/services/table';
import UiLanguageContext from 'contexts/uiLanguageContext';
import '@testing-library/jest-dom';
import { MemoryRouter } from 'react-router-dom';
import TableTreeSelection from './TableTreeSelection';
import { NavigationProvider } from 'contexts/navigationContext';

jest.mock('react-i18next', () => ({
    ...jest.requireActual('react-i18next'),
    useTranslation: () => {
        return {
            t: (str: string) => str,
            i18n: {
                changeLanguage: () => new Promise(() => ({})),
            },
        };
    },
}));

jest.mock('envVars', () => ({
    PxGrafUrl: 'pxGrafUrl.fi/',
    PublicUrl: 'publicUrl.fi/',
    BasePath: ''
}));

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useParams: () => {
        return {
            '*': 'foo/bar'
        };
    },
}));

const setLanguage = jest.fn();
const language = 'fi';

const setLanguageTab = jest.fn();
const languageTab = 'fi';
const availableUiLanguages = ['fi', 'en', 'sv'];

const uiContentLanguage = "fi";
const setUiContentLanguage = jest.fn();

const mockTableResult: ITableResult = {
    isLoading: false,
    isError: false,
    data: {
        headers: [],
        files: [
            {
                name: { 'fi': 'foo1', 'en': 'foo1', 'sv': 'foo1' },
                fileName: 'id1.px',
                lastUpdated: '1.1.2000',
                languages: ['fi', 'en', 'sv']
            },
            {
                name: { 'fi': 'foo2', 'en': 'foo2' },
                fileName: 'id2.px',
                lastUpdated: '1.1.2000',
                languages: ['fi', 'en']
            }
        ],
    }
}

jest.mock('api/services/table', () => ({
    ...jest.requireActual('api/services/table'),
    useTableQuery: () => {
        return mockTableResult;
    },
}));

describe('Rendering test', () => {
    it('renders correctly', () => {
        const { asFragment } = render(
            <MemoryRouter>
                <UiLanguageContext.Provider value={{ language, setLanguage, languageTab, setLanguageTab, availableUiLanguages, uiContentLanguage, setUiContentLanguage }}>
                    <NavigationProvider>
                        <TableTreeSelection />
                    </NavigationProvider>
                </UiLanguageContext.Provider>
            </MemoryRouter>
        );
        expect(asFragment()).toMatchSnapshot();
    });
});
