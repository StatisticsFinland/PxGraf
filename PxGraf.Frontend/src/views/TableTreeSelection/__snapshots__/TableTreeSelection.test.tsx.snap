// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rendering test renders correctly 1`] = `
<DocumentFragment>
  <div
    class="MuiContainer-root MuiContainer-maxWidthMd sc-ibQAlb iOxBWl css-xdmu94-MuiContainer-root"
  >
    <nav
      aria-labelledby="nested-list-subheader"
      class="MuiList-root MuiList-padding MuiList-subheader css-p64g7a-MuiList-root"
    >
      <div
        class="MuiListSubheader-root MuiListSubheader-gutters MuiListSubheader-sticky css-uob957-MuiListSubheader-root"
        id="nested-list-subheader"
      >
        <div
          class="sc-jGKxIK cLHFcM"
        >
          <div>
            tableSelect.title
          </div>
          <button
            aria-label="tooltip.open: tableSelect.title"
            class="sc-dAlyuH ghYgyh"
            id="mainContent"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-colorInfo MuiSvgIcon-fontSizeMedium css-1yq1soz-MuiSvgIcon-root"
              data-testid="InfoIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2m1 15h-2v-6h2zm0-8h-2V7h2z"
              />
            </svg>
          </button>
        </div>
      </div>
      <li
        class="MuiListItem-root MuiListItem-gutters MuiListItem-padding sc-iMTnTL gwgsRo css-1p823my-MuiListItem-root"
        id="id1"
      >
        <a
          class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters css-un12p1-MuiButtonBase-root-MuiListItemButton-root"
          href="/editor/id1/"
          tabindex="0"
        >
          <div
            class="MuiListItemIcon-root css-16myrmq-MuiListItemIcon-root"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="InsertDriveFileOutlinedIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M14 2H6c-1.1 0-1.99.9-1.99 2L4 20c0 1.1.89 2 1.99 2H18c1.1 0 2-.9 2-2V8zM6 20V4h7v5h5v11z"
              />
            </svg>
          </div>
          <div
            class="MuiListItemText-root MuiListItemText-multiline css-konndc-MuiListItemText-root"
          >
            <span
              class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
            >
              foo1 (FI, EN, SV)
            </span>
            <p
              class="MuiTypography-root MuiTypography-body2 MuiListItemText-secondary css-83ijpv-MuiTypography-root"
            >
              tableSelect.updated: 1.1.2000 klo 0.00.00
            </p>
          </div>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </a>
      </li>
      <hr
        class="MuiDivider-root MuiDivider-fullWidth css-9mgopn-MuiDivider-root"
      />
      <li
        class="MuiListItem-root MuiListItem-gutters MuiListItem-padding sc-iMTnTL gwgsRo css-1p823my-MuiListItem-root"
        id="id2"
      >
        <a
          class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters css-un12p1-MuiButtonBase-root-MuiListItemButton-root"
          href="/editor/id2/"
          tabindex="0"
        >
          <div
            class="MuiListItemIcon-root css-16myrmq-MuiListItemIcon-root"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="InsertDriveFileOutlinedIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M14 2H6c-1.1 0-1.99.9-1.99 2L4 20c0 1.1.89 2 1.99 2H18c1.1 0 2-.9 2-2V8zM6 20V4h7v5h5v11z"
              />
            </svg>
          </div>
          <div
            class="MuiListItemText-root MuiListItemText-multiline css-konndc-MuiListItemText-root"
          >
            <span
              class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
            >
              foo2 (FI, EN)
            </span>
            <p
              class="MuiTypography-root MuiTypography-body2 MuiListItemText-secondary css-83ijpv-MuiTypography-root"
            >
              tableSelect.updated: 1.1.2000 klo 0.00.00
            </p>
          </div>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </a>
      </li>
      <hr
        class="MuiDivider-root MuiDivider-fullWidth css-9mgopn-MuiDivider-root"
      />
    </nav>
  </div>
</DocumentFragment>
`;
